force-app/main/default/classes/Application_TDTM.cls

```diff
@@ -0,0 +1,77 @@
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+ * @author Salesforce.org
+ * @date 2021
+ * @group Applications
+ * @group-content ../../ApexDocContent/CourseEnrollments.htm
+ * @description This class prevents the application due date from being prior to the open date.
+ */
+public with sharing class Application_TDTM extends TDTM_Runnable {
+    /*******************************************************************************************************
+     * @description Stop an Application from being inserted when the Due Date is prior to the Open Date.
+     * @param listNew the list of Applications from trigger new.
+     * @param listOld the list of Applications from trigger old.
+     * @param triggerAction which trigger event (BeforeInsert, AfterInsert, etc.).
+     * @param objResult the describe for Application
+     * @return dmlWrapper.
+     ********************************************************************************************************/
+    public override DmlWrapper run(
+        List<SObject> newList,
+        List<SObject> oldList,
+        TDTM_Runnable.Action triggerAction,
+        Schema.DescribeSObjectResult objResult
+    ) {
+        if (triggerAction == TDTM_Runnable.Action.BeforeInsert || triggerAction == TDTM_Runnable.Action.BeforeUpdate) {
+            Map<Id, SObject> oldMap = (oldList != null) ? new Map<Id, SObject>(oldList) : new Map<Id, SObject>();
+
+            for (Application__c application : (List<Application__c>) newList) {
+                if (!application.isSet(Application__c.Open_Date__c) || !application.isSet(Application__c.Due_Date__c)) {
+                    continue;
+                }
+
+                if (triggerAction == TDTM_Runnable.Action.BeforeUpdate) {
+                    Application__c oldRecord = (Application__c) oldMap.get(application.Id);
+                    if (
+                        oldRecord.Open_Date__c == application.Open_Date__c &&
+                        oldRecord.Due_Date__c == application.Due_Date__c
+                    ) {
+                        continue;
+                    }
+                }
+                if (application.Open_Date__c >= application.Due_Date__c) {
+                    application.addError(Label.stgErrorAppDueDate);
+                }
+            }
+        }
+
+        return new DmlWrapper();
+    }
```

Comment:
1. The validation is only happening in the last if statement, all the other conditionals are excluding records from being validated. Therefore, if I was to split the validation logic into a separate method, I would only want to extract the last if statement, and I think that's overkill.
2. I took the `oldMap` and `Label.*` convention from other trigger files in EDA, and I think that being consistent would be more valuable in this case, since these are widely used conventions.

Reply:
1. I understand it might look like another way to write the same thing, but it is more readable if the code is extracted into a meaningful method with descriptive name. Moreover when/if we change the logic or extend behaviour, we will be extended the code while still keeping it readable if we start with clean base.
2. I understand the variable naming take take from the existing/old naming convention, thus I kept the `run()` method arguments the same. However, since this is a new trigger and new code, let's follow the best Apex coding style as of today (it may/will change in future and that is ok). Variable naming is very important to convey the behaviour as well reduce bugs. Previous NPSP/EDA naming convention contained data type in variables and I would like to reduce those occurrences.